x/vm/keeper/state_transition.go

```diff
@@ -86,47 +84,13 @@ func (k Keeper) GetHashFn(ctx sdk.Context) vm.GetHashFunc {
 			return common.Hash{}
 		}
 
-		switch {
-		case ctx.BlockHeight() == h:
-			// Case 1: The requested height matches the one from the context so we can retrieve the header
-			// hash directly from the context.
-			// Note: The headerHash is only set at begin block, it will be nil in case of a query context
+		if ctx.BlockHeight() == h {
 			headerHash := ctx.HeaderHash()
-			if len(headerHash) != 0 {
+			if len(headerHash) > 0 {
 				return common.BytesToHash(headerHash)
 			}
-
-			// only recompute the hash if not set (eg: checkTxState)
-			contextBlockHeader := ctx.BlockHeader()
-			header, err := cmttypes.HeaderFromProto(&contextBlockHeader)
-			if err != nil {
-				k.Logger(ctx).Error("failed to cast tendermint header from proto", "error", err)
-				return common.Hash{}
-			}
-
-			headerHash = header.Hash()
-			return common.BytesToHash(headerHash)
-
-		case ctx.BlockHeight() > h:
-			// Case 2: if the chain is not the current height we need to retrieve the hash from the store for the
-			// current chain epoch. This only applies if the current height is greater than the requested height.
-			histInfo, err := k.stakingKeeper.GetHistoricalInfo(ctx, h)
-			if err != nil {
-				k.Logger(ctx).Debug("error while getting historical info", "height", h, "error", err.Error())
-				return common.Hash{}
-			}
-
-			header, err := cmttypes.HeaderFromProto(&histInfo.Header)
-			if err != nil {
-				k.Logger(ctx).Error("failed to cast tendermint header from proto", "error", err)
-				return common.Hash{}
-			}
-
-			return common.BytesToHash(header.Hash())
-		default:
-			// Case 3: heights greater than the current one returns an empty hash.
-			return common.Hash{}
 		}
+		return common.BytesToHash(k.GetHeaderHash(ctx, h))
```

Comment:
in geth, it only ever services the hashes in the range of 256 blocks. otherwise it would return 0. is this assertion important to hold here? 

https://github.com/ethereum/go-ethereum/blob/5cc9137c9cec8a977b845a025d3deced9457fd48/core/vm/instructions.go#L431-L459

Reply:
I opened PR to implement EIP-2935, I think this PR can adapt to that later, see https://github.com/cosmos/evm/issues/406